pkg/eventrecorder/eventrecorder.go

```diff
@@ -11,55 +11,127 @@ import (
 
 	"github.com/filecoin-project/lassie/pkg/events"
 	"github.com/filecoin-project/lassie/pkg/types"
-	logging "github.com/ipfs/go-log/v2"
-	"github.com/multiformats/go-multicodec"
+	"github.com/ipfs/go-log/v2"
 )
 
-var HttpTimeout = 5 * time.Second
-var ParallelPosters = 5
+var logging = log.Logger("eventrecorder")
 
-var log = logging.Logger("eventrecorder")
+const httpTimeout = 5 * time.Second
+const parallelPosters = 5
 
-type EventRecorderConfig struct {
-	DisableIndexerEvents  bool
-	InstanceID            string
-	EndpointURL           string
-	EndpointAuthorization string
+type data struct {
+	source string
+	event  types.RetrievalEvent
 }
 
-// NewEventRecorder creates a new event recorder with the ID of this instance
-// and the URL to POST to
-func NewEventRecorder(ctx context.Context, cfg EventRecorderConfig) *EventRecorder {
-	er := &EventRecorder{
-		ctx,
-		make(chan report),
-		make(chan []report),
-		cfg,
-	}
-
-	go er.ingestReports()
-	for i := 0; i < ParallelPosters; i++ {
-		go er.postReports()
+// NewEventRecorderSubsciber returns a RetrievalEventSubscriber that POSTs
+// a batch of retrieval events to an event recorder API endpoint
+func NewEventRecorderSubscriber(
+	ctx context.Context,
+	instanceID string,
+	endpointURL string,
+	endpointAuthorization string,
+	disableIndexerEvents bool,
+) types.RetrievalEventSubscriber {
```

Comment:
I'm not greatly fussed on returning a struct vs a function, like @hannahhoward is, but I do recall making the decision to _not_ just return a function from this originally—it was based on the complexity which went past some arbitrary threshold where we needed too much state and having it in a struct was much cleaner than bundling it all in closures. The complexity has gone up here so my preference would be for structs rather than returning functions. You also get the benefit of being able to start/stop and inspect from outside when testing if you want.